core/src/main/java/org/apache/calcite/rel/rules/MinusToFilterRule.java

```diff
@@ -0,0 +1,141 @@
+public class MinusToFilterRule
+    extends RelRule<MinusToFilterRule.Config>
+    implements TransformationRule {
+
+  /** Creates an MinusToFilterRule. */
+  protected MinusToFilterRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Minus minus = call.rel(0);
+    if (minus.all || minus.getInputs().size() != 2) {
+      return;
+    }
+    final RelBuilder builder = call.builder();
+    final RelNode leftInput = call.rel(1);
+    final Filter rightInput = call.rel(2);
+
+    if (!RexUtil.isDeterministic(rightInput.getCondition())) {
+      return;
+    }
+
+    final Pair<RelNode, @Nullable RexNode> leftBase = getUnfilteredInputAndMergeCond(leftInput);
+    final Pair<RelNode, @Nullable RexNode> rightBase = getUnfilteredInputAndMergeCond(rightInput);
+    if (!leftBase.left.equals(rightBase.left)) {
+      return;
+    }
+
+    // Right input is Filter, right cond should be not null
+    assert rightBase.right != null;
+    final RexNode finalCond = leftBase.right != null
+        ? builder.and(leftBase.right, builder.not(rightBase.right))
+        : builder.not(rightBase.right);
+
+    builder.push(leftBase.left)
+        .filter(finalCond)
+        .distinct();
+
+    call.transformTo(builder.build());
+  }
+
+  /**
+   * This method iteratively traverses the input node, stripping away
+   * and extracting conditions from Filter operators at each level,
+   * ultimately merging all conditions into a composite condition
+   * while returning the first encountered non-Filter operator.
+   */
+  private static Pair<RelNode, @Nullable RexNode> getUnfilteredInputAndMergeCond(
+      final RelNode node) {
+    RexBuilder builder = node.getCluster().getRexBuilder();
+    RelNode stripped = node.stripped();
+    RexNode cond = null;
+    while (stripped instanceof Filter) {
+      Filter filter = (Filter) stripped;
+      if (cond == null) {
+        cond = filter.getCondition();
+      } else {
+        cond =
+            RexUtil.composeConjunction(builder, ImmutableList.of(cond, filter.getCondition()));
+      }
+      stripped = filter.getInput().stripped();
+    }
+    return Pair.of(stripped, cond);
+  }
+
+  /** Rule configuration. */
+  @Value.Immutable
+  public interface Config extends RelRule.Config {
+    Config DEFAULT = ImmutableMinusToFilterRule.Config.of()
+        .withOperandFor(Minus.class, RelNode.class, Filter.class);
+
+    @Override default MinusToFilterRule toRule() {
+      return new MinusToFilterRule(this);
+    }
+
+    /** Defines an operand tree for the given classes. */
+    default Config withOperandFor(Class<? extends Minus> minusClass,
+        Class<? extends RelNode> relNodeClass, Class<? extends Filter> filterClass) {
+      return withOperandSupplier(
+          b0 -> b0.operand(minusClass).inputs(
+              b1 -> b1.operand(relNodeClass).anyInputs(),
```

Comment:
We should have a test case covering RelNode and Filter. Now only include Filter and Filter.